projects: add dc2903a example project

mrcsosa opened the pull request:
Add example project for dc2903a/ltc2672 eval-board.

Signed-off-by: Marc Paolo Sosa <marcpaolo.sosa@analog.com>

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/analogdevicesinc/no-OS?pullRequest=1671) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/analogdevicesinc/no-OS?pullRequest=1671) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/analogdevicesinc/no-OS?pullRequest=1671) it.</sub>

rbolboac commented:
Hi @mrcsosa . It looks like the build for the project you added is failing. Could you please make the necessary changes?

rbolboac commented on projects/dc2903a/src.mk:
```diff
@@ -0,0 +1,36 @@
+
+SRCS += $(PROJECT)/src/platform/$(PLATFORM)/main.c
+
+#INCS += $(PROJECT)/src/common/app_config.h
```
this comment can be removed

rbolboac commented on projects/dc2903a/src/platform/maxim/main.c:
```diff
@@ -0,0 +1,90 @@
+#endif
+
+/***************************************************************************//**
+ * @brief Main function execution for STM32 platform.
```
update comment accordingly

rbolboac commented on projects/dc2903a/Makefile:
```diff
@@ -0,0 +1,9 @@
+# Select the example you want to enable by chossing y for enabling and n for disabling
+BASIC_EXAMPLE=y
+IIO_EXAMPLE=n
```
IIO related

rbolboac commented on projects/dc2903a/builds.json:
```diff
@@ -0,0 +1,10 @@
+		"basic_example_max32655": {
+			"flags": "BASIC_EXAMPLE=y IIO_EXAMPLE=n TARGET=max32655"
+		},
+		"iio_example_max32655": {
+			"flags": "BASIC_EXAMPLE=n IIO_EXAMPLE=y TARGET=max32655"
+		}
```
IIO related also this is why the build is failing

rbolboac commented on projects/dc2903a/src/examples/examples_src.mk:
```diff
@@ -0,0 +1,19 @@
+INCS += $(PROJECT)/src/examples/basic_example/basic_example.h
+endif
+
+ifeq (y,$(strip $(IIO_EXAMPLE)))
+TINYIIOD=y
+CFLAGS += -DIIO_EXAMPLE=1
+SRCS += $(PROJECT)/src/examples/iio_example/iio_example.c
+INCS += $(PROJECT)/src/examples/iio_example/iio_example.h
+endif
+
+ifeq (y,$(strip $(TINYIIOD)))
+SRC_DIRS += $(NO-OS)/iio/iio_app
+
+INCS += $(DRIVERS)/dac/ltc2672/iio_ltc2672.h
+SRCS += $(DRIVERS)/dac/ltc2672/iio_ltc2672.c
+endif
```
IIO related

rbolboac commented on projects/dc2903a/src/platform/maxim/main.c:
```diff
@@ -0,0 +1,90 @@
+	ret = basic_example_main();
+#endif
+
+#ifdef IIO_EXAMPLE
+	ret = iio_example_main();
+#endif
```
IIO related

rbolboac commented on projects/dc2903a/src/platform/platform_includes.h:
```diff
@@ -0,0 +1,53 @@
+#include "maxim/parameters.h"
+#endif
+
+#ifdef IIO_SUPPORT
+#include "iio_app.h"
+#endif
```
IIO related

rbolboac commented on projects/dc2903a/src/platform/maxim/main.c:
```diff
@@ -0,0 +1,90 @@
+#include "basic_example.h"
+#endif
+
+#ifdef IIO_EXAMPLE
+#include "iio_example.h"
+#endif
```
IIO related

rbolboac commented on projects/dc2903a/src/platform/maxim/main.c:
```diff
@@ -0,0 +1,90 @@
+#endif
+	return ret;
+
+#if (BASIC_EXAMPLE + IIO_EXAMPLE == 0)
+#error At least one example has to be selected using y value in Makefile.
+#elif (BASIC_EXAMPLE + IIO_EXAMPLE > 1)
```
IIO related

amiclaus reviewed (commented):
after doing the changes requested by @rbolboac please make sure you sign the CLA.

mrcsosa commented:
Changelogs ([cff7a64]):
Removed: "#INCS += $(PROJECT)/src/common/app_config.h" from projects/dc2903a/src.mk
Removed: IIO related code lines
Modified: STM32 -> maxim from projects/dc2903a/src.mk

others:
Signed the CLA

rbolboac commented on projects/dc2903a/src/common/common_data.c:
```diff
@@ -0,0 +1,72 @@
+/******************************************************************************/
+/********************** Macros and Constants Definitions **********************/
+/******************************************************************************/
+#ifdef BASIC_EXAMPLE
```
I know our examples used to contain this macro but I would remove it here because our new implementation for iio also needs this uart init parameter defined

rbolboac commented on projects/dc2903a/src/common/common_data.h:
```diff
@@ -0,0 +1,58 @@
+/******************************************************************************/
+/********************** Macros and Constants Definitions **********************/
+/******************************************************************************/
+#ifdef BASIC_EXAMPLE
```
I would remove the BASIC_EXAMPLE compiler switch here, see my previous remark

rbolboac commented on projects/dc2903a/src/platform/maxim/parameters.c:
```diff
@@ -0,0 +1,59 @@
+/******************************************************************************/
+/********************** Macros and Constants Definitions **********************/
+/******************************************************************************/
+#ifdef BASIC_EXAMPLE
```
I would remove the BASIC_EXAMPLE compiler switch, see my previous remark

rbolboac commented on projects/dc2903a/src/platform/maxim/parameters.h:
```diff
@@ -0,0 +1,75 @@
+/******************************************************************************/
+/********************** Macros and Constants Definitions **********************/
+/******************************************************************************/
+#ifdef BASIC_EXAMPLE
```
I would remove the BASIC_EXAMPLE compiler switch, see my previous remark

rbolboac commented on projects/dc2903a/src/examples/basic_example/basic_example.c:
```diff
@@ -0,0 +1,144 @@
+		ret = ltc2672_set_span_all_channels(ltc2672_desc,
+						    LTC2672_50VREF); // max current of 3.125mA
+		if (ret)
+			return ret;
```
ltc2672_init actually allocates some memory, before returning, ltc2672_remove should be called to avoid memory leaks

CiprianRegus replied:
It would be nice to do this since someone might reuse the code, but in our case, the program just ends if it gets out of the loop.

rbolboac commented on projects/dc2903a/src/examples/basic_example/basic_example.c:
```diff
@@ -0,0 +1,144 @@
+		for (i = 0; i < num_currents; i++) {
+			ret = ltc2672_set_current_all_channels(ltc2672_desc, currents_to_set[i]);
+			if (ret)
+				return ret;
```
same comment about memory leak as above, return only after ltc2672_remove is called

rbolboac commented on projects/dc2903a/src/examples/basic_example/basic_example.c:
```diff
@@ -0,0 +1,144 @@
+
+		ret = ltc2672_monitor_mux(ltc2672_desc, LTC2672_MUX_VREF);
+		if (ret)
+			return ret;
```
same comment about memory leak as above, return only after ltc2672_remove is called

rbolboac commented on projects/dc2903a/src/examples/basic_example/basic_example.c:
```diff
@@ -0,0 +1,144 @@
+
+		ret = ltc2672_chip_power_down(ltc2672_desc);
+		if (ret)
+			return ret;
```
same comment about memory leak as above, return only after ltc2672_remove is called

rbolboac commented on projects/dc2903a/src/examples/basic_example/basic_example.c:
```diff
@@ -0,0 +1,144 @@
+		ret = ltc2672_setup_toggle_channel(ltc2672_desc, LTC2672_DAC3,
+						   currents_to_toggle[0], currents_to_toggle[1]);
+		if (ret)
+			return ret;
```
same comment about memory leak as above, return only after ltc2672_remove is called

rbolboac commented on projects/dc2903a/src/examples/basic_example/basic_example.c:
```diff
@@ -0,0 +1,144 @@
+
+		ret = ltc2672_enable_toggle_channel(ltc2672_desc, 8);
+		if (ret)
+			return ret;
```
same comment about memory leak as above, return only after ltc2672_remove is called

rbolboac commented on projects/dc2903a/src/examples/basic_example/basic_example.c:
```diff
@@ -0,0 +1,144 @@
+
+			ret = ltc2672_global_toggle(ltc2672_desc, toggle_flag);
+			if (ret)
+				return ret;
```
same comment about memory leak as above, return only after ltc2672_remove is called

rbolboac commented on projects/dc2903a/src/examples/basic_example/basic_example.c:
```diff
@@ -0,0 +1,144 @@
+
+		ret = ltc2672_enable_toggle_channel(ltc2672_desc, 0);
+		if (ret)
+			return ret;
```
same comment about memory leak as above, return only after ltc2672_remove is called

rbolboac commented on projects/dc2903a/src/platform/maxim/main.c:
```diff
@@ -0,0 +1,80 @@
+
+	ret = no_os_uart_init(&ltc2672_uart_desc, &ltc2672_uart_ip);
+	if (ret)
+		return ret;
```
no_os_uart_init actually allocates some memory, before returning, no_os_uart_remove should be called to avoid memory leaks

CiprianRegus commented on projects/dc2903a/src/common/common_data.c:
```diff
@@ -0,0 +1,72 @@
+	.extra = SPI_EXTRA,
+};
+
+struct ltc2672_init_param  ltc2672_ip = {
```
remove the double spacing

CiprianRegus commented on projects/dc2903a/src/examples/basic_example/basic_example.c:
```diff
@@ -0,0 +1,144 @@
+	/* Continuously configure the part and measure pins with multimeter for verification */
+	while (1) {
+		ret = ltc2672_set_span_all_channels(ltc2672_desc,
+						    LTC2672_50VREF); // max current of 3.125mA
```
The comment style should be `/**/`

mrcsosa commented:
Hello, I have updated the files with necessary changes, although I am not sure if I added the no_os_remove and ltc2672_remove correctly.
